nutils/export.py

```diff
@@ -0,0 +1,153 @@
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+from . import config, log, core, util
+import contextlib, numpy
+
+@contextlib.contextmanager
+@util.positional_only('name')
+def mplfigure(*args, **kwargs):
+  '''Matplotlib figure context, convenience function.
+
+  Returns a figure object suitable for `object-oriented plotting`_. Upon exit
+  the result is saved using the agg-backend in all formats configured via
+  config.imagetype, and the resulting filenames written to log.
+
+  .. _`object-oriented plotting`: https://matplotlib.org/gallery/api/agg_oo_sgskip.html
+
+  Args
+  ----
+  name : :class:`str`
+      The filename (without extension) of the resulting figure(s)
+  ... :
+      Keyword arguments are passed on unchanged to the constructor of the
+      ``Figure`` object.
+  '''
+
+  name, = args
+  formats = config.imagetype.split(',')
+  with log.context(name):
+    import matplotlib.backends.backend_agg
+    fig = matplotlib.figure.Figure(**kwargs)
+    matplotlib.backends.backend_agg.FigureCanvas(fig) # sets reference via fig.set_canvas
+    yield fig
+    for fmt in formats:
+      with log.context(fmt), core.open_in_outdir(name+'.'+fmt, 'wb') as f:
+        fig.savefig(f, format=fmt)
+    fig.set_canvas(None) # break circular reference
+  log.user(', '.join(name+'.'+fmt for fmt in formats))
+
+@util.positional_only('path', 'tri', 'x')
+def vtk(*args, **kwargs):
+  '''Export data to a VTK file.
+
+  This method provides a simple interface to the `VTK file format`_ with a
+  number of important restrictions:
+
+  *   Simplex-only. This makes it possible to define the mesh by a combination
+      of vertex coordinates and a connectivity table.
+  *   Legacy mode. The newer XML based format is more complex and does not
+      provide benefits within the constraints set by this method.
+  *   Binary mode. This allows for direct output of binary data, which aids
+      both speed and file size.
+
+  Beyond the mandatory file name, connectivity table, and vertex coordinates,
+  any additional data sets can be provided as keyword arguments, where the keys
+  are the names by which the arrays are stored. The data can be either vertex
+  or point data, with the distinction made based on the length of the array.
+
+  .. _`VTK file format`: https://www.vtk.org/VTK/img/file-formats.pdf
+
+  Args
+  ----
+  name : :class:`str`
+    Destination file name (without vtk extension).
+  tri : :class:`int` array
+    Triangulation.
+  x : :class:`float` array
+    Vertex coordinates.
+  ... :
+    Cell and/or point data
+  '''
+
+  assert len(args) == 3, 'expected 3 positional arguments: name, tri, x'
```

Comment:
Again, `name` is used here, but called `path` in the signature.

Reply:
I prefer an `if ...: raise TypeError('...')` here. Maybe the check should be the responsibility of `util.positional_only`.